src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMap.java

```diff
@@ -99,4 +108,40 @@ static synchronized void removeJobWithBuildConfigUid(String uid) {
         buildConfigToJobMap.remove(uid);
     }
 
+    /**
+     * Reconciles job runs and OpenShift builds
+     *
+     * Deletes all job runs that do not have an associated build in OpenShift
+     */
+    private static synchronized void reconcileRunsAndBuilds() {
+        logger.info("Reconciling job runs with builds");
+
+        for (Map.Entry<String, WorkflowJob> entry : buildConfigToJobMap.entrySet()) {
+            WorkflowJob job = entry.getValue();
+            BuildConfigProjectProperty bcpp = job.getProperty(BuildConfigProjectProperty.class);
+            BuildList buildList = getAuthenticatedOpenShiftClient().builds()
+                .inNamespace(bcpp.getNamespace()).withLabel("buildconfig=" + bcpp.getName()).list();
+
+            logger.info("Checking runs for BuildConfig " + bcpp.getNamespace() + "/" + bcpp.getName());
+
+            for (WorkflowRun run : job.getBuilds()) {
+              boolean found = false;
+              BuildCause cause = run.getCause(BuildCause.class);
+              for (Build build : buildList.getItems()) {
+                if (cause != null && cause.getUid().equals(build.getMetadata().getUid())) {
+                  found = true;
+                  break;
+                }
+              }
+              if (!found) {
+                try {
+                  logger.info("Deleting run: " + run.toString());
+                  run.delete();
```

Comment:
should probably call your deleteRun() helper instead so we only have a single codepath that deletes run (especially in case it starts getting messier w/ us having to do config save operations or something else to sync settings)

Reply:
I am unable to use the deleteRun method as it takes a job and a build, and since there is no build associated with the run because it has already been deleted ....

Reply:
touche